### DISCUSSION TITLE
How to raise a BadDataException when a column is not compliant ?

### DISCUSSION THREAD

**Discussion by LLFT** [JoshClose/CsvHelper#2137] [Q&A]
Hello,

How do I check that the first column of the CSV consists of 12 characters and raise a BadDataFound?

Example of CSV:
```
Example of CSV:
701959237444,Greyson
687412989760,Henry
65577910137,Valk
700266014739,Hann
716621078785,Royal
463239457655,Tae
580391672789,Hugo
```

An extract of my code
```csharp
using (StreamReader reader = new(_pathCSV, Encoding.UTF8, true))
{
	List<string> bads = new List<string>();
	bool isRecordBad = false;
	// On Parse le fichier CSV
	var config = new CsvConfiguration(CultureInfo.InvariantCulture)
	{
		BadDataFound = context =>
		{
			isRecordBad = true;
			bads.Add(context.Context.Parser.Row.ToString());
		},                            
	};
	using (CsvReader csv = new(reader, config))
	{
		recordsImport = csv.GetRecords<DataByCategorieImport>().ToList();
	}
	if (isRecordBad)
	{
		foreach (string bad in bads)
		{
			if (bad != null)
			{
				MessageBox.Show($"The file is not in conformity on line {bad}.", "Treatment stopped", MessageBoxButtons.OK, MessageBoxIcon.Error); ;
			}
		}
	}
}
```

**Comment by A9G-Data-Droid** [comment]
The `BadDataException` feature only throws when the file fails to conform to the CSV specification. What you are talking about is business logic. You want to throw your own exception for your own business logic. In this case I would add the logic to the setter of the property you are loading because the CSV reader will use your setters. This enforces your business logic all the time for your object regardless of how data is loaded into the object (CSV\JSON\Manually). Now it's not tied to this method of loading the data!

When you follow my example the exception you receive will be `CsvHelper.ReaderException: An unexpected error occurred.` which will contain the exception you threw as an inner exception, in my example it looks like `---> System.ApplicationException: ID must be 12 characters.`

EXAMPLE:
```
    public class DataByCategorieImport
    {
        private string? _id;

        [Index(0)]
        public string? ID
        {
            get => _id;
            set
            {
                if (value.Length != 12) throw new ApplicationException("ID must be 12 characters.");
                _id= value;
            }
        }
...
```
